src/shared/copy.c

```diff
@@ -1162,13 +1164,22 @@ static int fd_copy_directory(
 
         fdt = xopenat_lock_full(dt, to,
                                 O_RDONLY|O_DIRECTORY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW|(exists ? 0 : O_CREAT|O_EXCL),
-                                (copy_flags & COPY_MAC_CREATE ? XO_LABEL : 0)|(set_contains(subvolumes, st) ? XO_SUBVOLUME : 0),
+                                (copy_flags & COPY_MAC_CREATE ? XO_LABEL : 0)|(hashmap_contains(subvolumes, st) ? XO_SUBVOLUME : 0),
                                 st->st_mode & 07777,
                                 copy_flags & COPY_LOCK_BSD ? LOCK_BSD : LOCK_NONE,
                                 LOCK_EX);
         if (fdt < 0)
                 return fdt;
 
+        if (hashmap_contains(subvolumes, st)) {
+                subvolume_flags = PTR_TO_UINT(hashmap_get(subvolumes, st));
+                if (subvolume_flags & BTRFS_SUBVOL_NODATACOW) {
+                        r = btrfs_subvol_set_nodatacow_fd(fdt, true);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to make subvolume '%s' nodatacow: %m", to);
+                }
+        }
```

Comment:
Please rework this to pass `XO_NOCOW` to the `xopenat_lock_full()` above if the `NODATACOW` flag is set.

You can also remove `btrfs_subvol_get_nodatacow_fd()` and `btrfs_subvol_set_nodatacow_fd()` then.